**0. Your name, roll number and email-id**

**Answer in the textbox below**


Ankit Dixit , MT19AI015, ankit.1@iitj.ac.in

**This assignment pertains to concrete architecture for deliberatve agents, and BDI modeling. Our agent wants to either play tennis or watch a movie. We model the agent as an B-D-I agent and implement it with Bayesian Network. Please refer to Module 06-05 for details.**

Install libraries

In [ ]:
!pip install pomegranate

     |████████████████████████████████| 4.5MB 2.8MB/s 
  Created wheel for pomegranate: filename=pomegranate-0.13.5-cp36-cp36m-linux_x86_64.whl size=15238396 sha256=e9d1cf8835148cee98ef709936e4df4db37ae270151a934f86fe9dd09b279d67
  Stored in directory: /root/.cache/pip/wheels/99/d9/0a/0758302ccad41a61aa08ae4985b84d4cab5086e99c92a68560
Successfully built pomegranate


**1:** Create a Bayesian Network as shown in slide number 4 to model the belief of the agent. Use the same prior and contitional probabilities. Initialize the network.

**Put your code in the space below**

In [ ]:
from pomegranate import *
import numpy as np
#M --> Movie ,A--> Availble ,NA --> Not Availble , S--> Sunny , C --> Cloudy , NM --> Not movie , NT --> Not Tennis 
weather=DiscreteDistribution({'S':0.7,'C':0.3})
Partner=DiscreteDistribution({'A':0.5,'NA':0.5})

Movie= ConditionalProbabilityTable([['S','A','M',0.01],
                                    ['S','A','NM',0.99],
                                    ['S','NA','M',0.60],
                                    ['S','NA','NM',0.40],
                                    ['C','A','M',0.10],
                                    ['C','A','NM',0.90],
                                    ['C','NA','M',0.90],
                                    ['C','NA','NM',0.10]],[weather,Partner])
Tennis=ConditionalProbabilityTable([['S','A','T',0.95],
                                    ['S','A','NT',0.05],
                                    ['S','NA','T',0.10],
                                    ['S','NA','NT',0.90],
                                    ['C','A','T',0.10],
                                    ['C','A','NT',0.90],
                                    ['C','NA','T',0.01],
                                    ['C','NA','NT',0.99]],[weather,Partner])

S1=Node(weather,name="weather")
S2=Node(Partner,name="Partner")
S3=Node(Movie,name="Movie")
S4=Node(Tennis,name="Tennis")

In [ ]:
# Creating Bayesian network 
network=BayesianNetwork("Network to estimate the movie or tennis")
network.add_states(S1,S2,S3,S4)
network.add_edge(S1,S3)
network.add_edge(S1,S4)
network.add_edge(S2,S3)
network.add_edge(S2,S4)
network.bake()
#print(network)
#network.plot()


In [ ]:
observations = { }
beliefs = map( str, network.predict_proba( observations ) )
print("\n".join( "{}\t\t{}".format( state.name, belief ) for state, belief in zip( network.states, beliefs ) ))

weather		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "S" : 0.6999999999999998,
            "C" : 0.3000000000000002
        }
    ],
    "frozen" : false
}
Partner		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "A" : 0.5,
            "NA" : 0.5
        }
    ],
    "frozen" : false
}
Movie		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "NM" : 0.6365,
            "M" : 0.36350000000000016
        }
    ],
    "frozen" : false
}
Tennis		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "NT" : 0.6160000000000001,
            "T" : 0.384
        }
    ],
    "frozen" : false
}


**2.** What are the marginal probabilities P(M=y) and P(T=y)? Verify against the values given on the slide.

What are the desires for the agent? Assume a threshold value of **0.3** for feasible desires. Which of the desires are feasible?

**Answer in the text-box below**


The marginal Probabilities for P(M=y) is 0.3635 and P(T=y) is 0.384 , The desires for the agent is choose {M,T} 
For a thresold value of  0.3 both the desires are satisfying but the agent chooses one with higher marginal probability , so P(T=y)= 0.384 are feasible .

**3:** The agent wants to pursue a feasible desire with higher marginal probability (playing tennis). That becomes intention and the agent commits to it.

The agent needs to act to realize the intention. He looks at the weather and perceives it to be sunny.

Instantiate "Weather" node (W=s). Update the belief of the agent with this perception. 

**Put your code in the box below**

In [ ]:
#Instantiating "weather" node (W=s)
observations = { 'weather':'S'}
beliefs = map( str, network.predict_proba( observations ) )
print("\n".join( "{}\t\t{}".format( state.name, belief ) for state, belief in zip( network.states, beliefs ) ))

weather		S
Partner		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "A" : 0.5,
            "NA" : 0.5
        }
    ],
    "frozen" : false
}
Movie		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "NM" : 0.695,
            "M" : 0.30500000000000005
        }
    ],
    "frozen" : false
}
Tennis		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "NT" : 0.47500000000000003,
            "T" : 0.525
        }
    ],
    "frozen" : false
}


**4.** (a) What are the marginal probabilities P(M=y) and P(T=y) now? Which of the desires are feasible now? If there are no feasible desires, *stop here*.

(b) If the current intention remains feasible, the agent pursues it. Otherwise, the agent commits to another feasible desire, which will becomes intention. What will be the intention now?

**Answer in the text-box below**

(a).The marginal probabilites now is P(M=y) = 0.305 and P(T=y) is 0.525 .
For a thersold value of 0.3 both are satisfying but agian the agent chooses with higher marginal probabilites so it chooses P(T=y) with 0.525 , 
(b).The agent persues the current intention i.e. Tennis.
So the intenstion now is play Tennis.

**5.** The pro-active agent acts further on it's intention. It seeks a partner, but *fails* to get one.

Instantiate "Partner" node (P=n). Update the belief of the agent.

**Put your code in the box below.**


In [ ]:
#Instantiating "partner " Node (P=NA)
observations = { 'weather':'S','Partner':'NA'}
beliefs = map( str, network.predict_proba( observations ) )
print("\n".join( "{}\t\t{}".format( state.name, belief ) for state, belief in zip( network.states, beliefs ) ))

weather		S
Partner		NA
Movie		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "NM" : 0.4000000000000001,
            "M" : 0.5999999999999999
        }
    ],
    "frozen" : false
}
Tennis		{
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "NT" : 0.8999999999999999,
            "T" : 0.10000000000000014
        }
    ],
    "frozen" : false
}


**6.** 
(a) What are the marginal probabilities P(M=y) and P(T=y) now? What are the feasible desires? If the current intention is still feasible, *stop here*.

(b) Withdraw commitment to the intention, and choose a feasible desire as commitment. What is the commitment now?

**Answer in the text-box below**

The marginal probabilities for P(M=y) is 0.599 and P(T=y) is 0.10 , In this case the marginal probability for tennis is less than the marginal probability of Movie and also less than the thersold value 0.3, so feasible desire is Movie. 
(b) So the agent withdraw its commitment i.e. Tennis , and Chooses feasible desire Movie as commitment , The commitment now is Movie.

**7.** Is this agent a *learning* agent? Justify your answer.

If it is not a learning agent, how can you make it one?

**Answer in the text-box below**


No the agent is not learning agent , because the enviorment is dynamic and it does not have an ability to adapt the changes ,this can be made as learning agent by providing explicit representation of its enviroment,this logical representation captures the real world enviorment in some set of predicates(first order logic), and by providing some perfomace measure which tells which actions leads to new experiences , in addition to that by providing some learning element which tells the agent to which actions might lead to bad experiences.
It is also done with reinforcement learning in which each action leads to some reward or penalty . 

**8. DISCUSSIONS**

Analyze the steps and the results above. Summarize in plain English how the belief, desire and commitment of agent changed with percepts. 

Make any salient observations that you think suitable.

**Answer in the text-box below**

In our example the belief depends upon the percepts it changes in second cycle when the actions in first cycle fails , if percepts changes (i.e. Partner not availble) we instatiate that node(Partner =NA) and hence the belief revision takes place based on current beilef and perception.
Desire : The desire is not gurantee to mature like in our case if we are unable to get the partner in that case our percept changes and hence the desire change from tennis to movie.
Commitment :  Agent commitment based on the feasible desires , first we explore only the weather node and based on that we obtain feasible desire , but after the exploration other parent node (Partner), we observe the change in feasible desires(infeasible Tennis) and hence we commit to other option i.e. Movie. 

**Congratulations!** Your assignment is complete.

**Send review invitation to** hiranmay@gmail.com